feat(router): auto-unwrap default exports when lazy loading

alxhub opened the pull request:
When using `loadChildren` or `loadComponent`, a common pattern is to pass a function that returns a `Promise` from a dynamic import:

```typescript
{
  path: 'lazy',
  loadComponent: () => import('./lazy-file').then(m => m.LazyCmp),
}
```

The `.then` part of the expression selects the particular exported component symbol from the dynamically imported ES module.

ES modules can have a "default export", created with the `export default` modifier:

```typescript
@Component({...})
export default class LazyCmp { ... }
```

This default export is made available to dynamic imports under the well- known key of `'default'`, per the ES module spec:
https://tc39.es/ecma262/#table-export-forms-mapping-to-exportentry-records

This commit adds a feature to the router to automatically dereference such default exports. With this logic, when `export default` is used, a `.then` operation to select the particular exported symbol is no longer required:

```typescript
{
  path: 'lazy',
  loadComponent: () => import('./lazy-file'),
}
```

The above `loadComponent` operation will automatically use the `default` export of the `lazy-file` ES module.

This functionality works for `loadChildren` as well.

alxhub committed:
feat(router): auto-unwrap default exports when lazy loading

When using `loadChildren` or `loadComponent`, a common pattern is to pass
a function that returns a `Promise` from a dynamic import:

```typescript
{
  path: 'lazy',
  loadComponent: () => import('./lazy-file').then(m => m.LazyCmp),
}
```

The `.then` part of the expression selects the particular exported
component symbol from the dynamically imported ES module.

ES modules can have a "default export", created with the `export default`
modifier:

```typescript
@Component({...})
export default class LazyCmp { ... }
```

This default export is made available to dynamic imports under the well-
known key of `'default'`, per the ES module spec:
https://tc39.es/ecma262/#table-export-forms-mapping-to-exportentry-records

This commit adds a feature to the router to automatically dereference such
default exports. With this logic, when `export default` is used, a `.then`
operation to select the particular exported symbol is no longer required:

```typescript
{
  path: 'lazy',
  loadComponent: () => import('./lazy-file'),
}
```

The above `loadComponent` operation will automatically use the `default`
export of the `lazy-file` ES module.

This functionality works for `loadChildren` as well.

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

(just one comment about the DX for error cases)

atscott commented on packages/router/test/standalone.spec.ts:
```diff
@@ -340,6 +339,35 @@ describe('standalone in Router API', () => {
+      expect(root.nativeElement.innerHTML).toContain('default exported');
+    });
+
+    it('should work for loadChildren', async () => {
```
TypeScript will fail in both cases: 

If there are no exports, the TypeScript compiler will error with `File 'the/imported/thing' is not a module.`

If there is no default export, it would fail the `DefaultExport<T>` type.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

AndrewKushnir commented:
FYI, it looks like this change would require an update to the local patches in g3, adding the "blocked" label for now...

alxhub committed:
refactor(router): organize deprecated loadChildren code

This commit breaks out the code in @angular/router needed to support the
deprecated `loadChildren` signature into a separate file, which reduces the
opportunity for conflicts when patching that behavior in g3.

alxhub commented:
Caretaker: a g3 patch (cl/478602165) is required

atscott commented on goldens/public-api/router/index.md:
```diff
@@ -357,7 +357,7 @@ export interface IsActiveMatchOptions {
 
 // @public
-export type LoadChildren = LoadChildrenCallback;
+export type LoadChildren = LoadChildrenCallback | ɵDeprecatedLoadChildren;
```
`ɵDeprecatedLoadChildren` now appears in the public API even though it's not exported and is unused externally. What about moving the `LoadChildren` type to its own file that we patch in g3 to add `ɵDeprecatedLoadChildren`?

alxhub replied:
Yep - that's fine. The goal is to not need to touch the definition of `LoadChildren` at all in the patch file - that's what leads to brittle patch files.

atscott commented on packages/router/src/deprecated_load_children.ts:
```diff
@@ -0,0 +1,27 @@
+export function deprecatedLoadChildrenString(
+    injector: Injector, loadChildren: string): Observable<NgModuleFactory<any>> {
+  // return from(injector.get(NgModuleFactoryLoader).load(loadChildren));
+  throw new Error('Assertion failure - loadChildren strings not supported');
```
Isn't this error now non-tree shakeable?

atscott commented on packages/router/src/router_config_loader.ts:
```diff
@@ -122,6 +123,9 @@ export class RouterConfigLoader {
   private loadModuleFactoryOrRoutes(loadChildren: LoadChildren):
       Observable<NgModuleFactory<any>|Routes> {
+    if (typeof loadChildren === 'string') {
+      return deprecatedLoadChildrenString(this.injector, loadChildren);
```
What if we unconditionally call `deprecatedLoadChildrenString` from here and make it an empty function externally? We at least have a single file we can patch in g3 then.

alxhub replied:
We could restructure it so that if `deprecatedLoadChildrenString` returns _something_, then we return that, yeah.

alxhub replied:
This doesn't work, because the `typeof` needs to be performed within `loadModuleFactoryOrRoutes` to narrow the type of `loadChildren` for the rest of the function.

AndrewKushnir commented:
This PR was merged into the repository by commit da58801f95c66c201e332189af25702bdd722f3f.
